File: backend/app/schemas/proto_assets.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from backend.app.domain.proto_asset import (
    ProtoAsset,
    ProtoDecodeResult,
    ProtoEncodeResult,
    ProtoEnum,
    ProtoEnumValue,
    ProtoField,
    ProtoMessage,
    ProtoMethod,
    ProtoService,
)

SHA256_PATTERN = r"^[0-9a-f]{64}$"
MESSAGE_TYPE_PATTERN = r"^[A-Za-z_][A-Za-z0-9_.]{0,254}$"


class ProtoImportRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    source_path: str = Field(min_length=1, max_length=32767)


class ProtoFieldResponse(BaseModel):
    name: str
    number: int
    type: str
    label: str
    type_name: str | None

    @classmethod
    def from_domain(cls, field: ProtoField) -> "ProtoFieldResponse":
        return cls(**{name: getattr(field, name) for name in cls.model_fields})


class ProtoMessageResponse(BaseModel):
    name: str
    full_name: str
    fields: list[ProtoFieldResponse]

    @classmethod
    def from_domain(cls, item: ProtoMessage) -> "ProtoMessageResponse":
        return cls(
            name=item.name,
            full_name=item.full_name,
            fields=[ProtoFieldResponse.from_domain(field) for field in item.fields],
        )


class ProtoEnumValueResponse(BaseModel):
    name: str
    number: int

    @classmethod
    def from_domain(cls, item: ProtoEnumValue) -> "ProtoEnumValueResponse":
        return cls(name=item.name, number=item.number)


class ProtoEnumResponse(BaseModel):
    name: str
    full_name: str
    values: list[ProtoEnumValueResponse]

    @classmethod
    def from_domain(cls, item: ProtoEnum) -> "ProtoEnumResponse":
        return cls(
            name=item.name,
            full_name=item.full_name,
            values=[ProtoEnumValueResponse.from_domain(value) for value in item.values],
        )


class ProtoMethodResponse(BaseModel):
    name: str
    input_type: str
    output_type: str
    client_streaming: bool
    server_streaming: bool

    @classmethod
    def from_domain(cls, item: ProtoMethod) -> "ProtoMethodResponse":
        return cls(**{name: getattr(item, name) for name in cls.model_fields})


class ProtoServiceResponse(BaseModel):
    name: str
    full_name: str
    methods: list[ProtoMethodResponse]

    @classmethod
    def from_domain(cls, item: ProtoService) -> "ProtoServiceResponse":
        return cls(
            name=item.name,
            full_name=item.full_name,
            methods=[ProtoMethodResponse.from_domain(method) for method in item.methods],
        )


class ProtoAssetResponse(BaseModel):
    id: str
    workspace_id: str
    name: str
    relative_path: str
    sha256: str
    size_bytes: int
    packages: list[str]
    messages: list[ProtoMessageResponse]
    enums: list[ProtoEnumResponse]
    services: list[ProtoServiceResponse]
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_domain(cls, asset: ProtoAsset) -> "ProtoAssetResponse":
        return cls(
            id=asset.id,
            workspace_id=asset.workspace_id,
            name=asset.name,
            relative_path=asset.relative_path,
            sha256=asset.sha256,
            size_bytes=asset.size_bytes,
            packages=list(asset.packages),
            messages=[ProtoMessageResponse.from_domain(item) for item in asset.messages],
            enums=[ProtoEnumResponse.from_domain(item) for item in asset.enums],
            services=[ProtoServiceResponse.from_domain(item) for item in asset.services],
            created_at=asset.created_at,
            updated_at=asset.updated_at,
        )


class ProtoEncodeRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_sha256: str = Field(pattern=SHA256_PATTERN)
    message_type: str = Field(min_length=1, max_length=255, pattern=MESSAGE_TYPE_PATTERN)
    payload: dict[str, Any]


class ProtoDecodeRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_sha256: str = Field(pattern=SHA256_PATTERN)
    message_type: str = Field(min_length=1, max_length=255, pattern=MESSAGE_TYPE_PATTERN)
    data_base64: str = Field(max_length=2_796_212)


class ProtoEncodeResponse(BaseModel):
    data_base64: str
    size_bytes: int

    @classmethod
    def from_domain(cls, result: ProtoEncodeResult) -> "ProtoEncodeResponse":
        return cls(data_base64=result.data_base64, size_bytes=result.size_bytes)


class ProtoDecodeResponse(BaseModel):
    payload: dict[str, Any]
    size_bytes: int

    @classmethod
    def from_domain(cls, result: ProtoDecodeResult) -> "ProtoDecodeResponse":
        return cls(payload=result.payload, size_bytes=result.size_bytes)

File: frontend/src/api/proto-assets.ts
import { resolveBackendConnection } from "./backend-connection";
import { ApiClient, ApiClientError } from "./client";

export interface ProtoField {
  name: string;
  number: number;
  type: string;
  label: string;
  type_name: string | null;
}

export interface ProtoMessage {
  name: string;
  full_name: string;
  fields: ProtoField[];
}

export interface ProtoEnum {
  name: string;
  full_name: string;
  values: Array<{ name: string; number: number }>;
}

export interface ProtoMethod {
  name: string;
  input_type: string;
  output_type: string;
  client_streaming: boolean;
  server_streaming: boolean;
}

export interface ProtoService {
  name: string;
  full_name: string;
  methods: ProtoMethod[];
}

export interface ProtoAsset {
  id: string;
  workspace_id: string;
  name: string;
  relative_path: string;
  sha256: string;
  size_bytes: number;
  packages: string[];
  messages: ProtoMessage[];
  enums: ProtoEnum[];
  services: ProtoService[];
  created_at: string;
  updated_at: string;
}

export interface ProtoEncodeInput {
  expected_sha256: string;
  message_type: string;
  payload: Record<string, unknown>;
}

export interface ProtoEncodeResult {
  data_base64: string;
  size_bytes: number;
}

export interface ProtoDecodeInput {
  expected_sha256: string;
  message_type: string;
  data_base64: string;
}

export interface ProtoDecodeResult {
  payload: Record<string, unknown>;
  size_bytes: number;
}

function isStringArray(value: unknown): value is string[] {
  return Array.isArray(value) && value.every((item) => typeof item === "string");
}

function isProtoField(value: unknown): value is ProtoField {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Partial<ProtoField>;
  return (
    typeof item.name === "string" &&
    Number.isInteger(item.number) &&
    typeof item.type === "string" &&
    typeof item.label === "string" &&
    (item.type_name === null || typeof item.type_name === "string")
  );
}

function isProtoMessage(value: unknown): value is ProtoMessage {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Partial<ProtoMessage>;
  return (
    typeof item.name === "string" &&
    typeof item.full_name === "string" &&
    Array.isArray(item.fields) &&
    item.fields.every(isProtoField)
  );
}

function isProtoEnum(value: unknown): value is ProtoEnum {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Partial<ProtoEnum>;
  return (
    typeof item.name === "string" &&
    typeof item.full_name === "string" &&
    Array.isArray(item.values) &&
    item.values.every(
      (entry) =>
        typeof entry === "object" &&
        entry !== null &&
        typeof entry.name === "string" &&
        Number.isInteger(entry.number),
    )
  );
}

function isProtoService(value: unknown): value is ProtoService {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Partial<ProtoService>;
  return (
    typeof item.name === "string" &&
    typeof item.full_name === "string" &&
    Array.isArray(item.methods) &&
    item.methods.every(
      (method) =>
        typeof method === "object" &&
        method !== null &&
        typeof method.name === "string" &&
        typeof method.input_type === "string" &&
        typeof method.output_type === "string" &&
        typeof method.client_streaming === "boolean" &&
        typeof method.server_streaming === "boolean",
    )
  );
}

function isProtoAsset(value: unknown): value is ProtoAsset {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Partial<ProtoAsset>;
  return (
    typeof item.id === "string" &&
    typeof item.workspace_id === "string" &&
    typeof item.name === "string" &&
    typeof item.relative_path === "string" &&
    typeof item.sha256 === "string" &&
    /^[0-9a-f]{64}$/.test(item.sha256) &&
    Number.isInteger(item.size_bytes) &&
    isStringArray(item.packages) &&
    Array.isArray(item.messages) &&
    item.messages.every(isProtoMessage) &&
    Array.isArray(item.enums) &&
    item.enums.every(isProtoEnum) &&
    Array.isArray(item.services) &&
    item.services.every(isProtoService) &&
    typeof item.created_at === "string" &&
    typeof item.updated_at === "string"
  );
}

function asset(value: unknown): ProtoAsset {
  if (!isProtoAsset(value)) {
    throw new ApiClientError("后端 Proto 资产响应格式不正确。", {
      code: "INVALID_RESPONSE",
    });
  }
  return value;
}

async function client(): Promise<ApiClient> {
  const connection = await resolveBackendConnection();
  return new ApiClient(connection.baseUrl, connection.token);
}

export async function listProtoAssets(workspaceId: string): Promise<ProtoAsset[]> {
  const response = await (await client()).get<unknown>(
    `/api/workspaces/${workspaceId}/proto-assets`,
  );
  if (!Array.isArray(response) || !response.every(isProtoAsset)) {
    throw new ApiClientError("后端 Proto 资产列表响应格式不正确。", {
      code: "INVALID_RESPONSE",
    });
  }
  return response;
}

export async function importProtoAsset(
  workspaceId: string,
  sourcePath: string,
): Promise<ProtoAsset> {
  return asset(
    await (await client()).post<unknown>(`/api/workspaces/${workspaceId}/proto-assets`, {
      source_path: sourcePath,
    }),
  );
}

export async function encodeProtoMessage(
  workspaceId: string,
  assetId: string,
  input: ProtoEncodeInput,
): Promise<ProtoEncodeResult> {
  const response = await (await client()).post<unknown>(
    `/api/workspaces/${workspaceId}/proto-assets/${assetId}/encode`,
    input,
  );
  if (
    typeof response !== "object" ||
    response === null ||
    typeof (response as Partial<ProtoEncodeResult>).data_base64 !== "string" ||
    !Number.isInteger((response as Partial<ProtoEncodeResult>).size_bytes)
  ) {
    throw new ApiClientError("后端 Protobuf 编码响应格式不正确。", {
      code: "INVALID_RESPONSE",
    });
  }
  return response as ProtoEncodeResult;
}

export async function decodeProtoMessage(
  workspaceId: string,
  assetId: string,
  input: ProtoDecodeInput,
): Promise<ProtoDecodeResult> {
  const response = await (await client()).post<unknown>(
    `/api/workspaces/${workspaceId}/proto-assets/${assetId}/decode`,
    input,
  );
  if (
    typeof response !== "object" ||
    response === null ||
    typeof (response as Partial<ProtoDecodeResult>).payload !== "object" ||
    (response as Partial<ProtoDecodeResult>).payload === null ||
    Array.isArray((response as Partial<ProtoDecodeResult>).payload) ||
    !Number.isInteger((response as Partial<ProtoDecodeResult>).size_bytes)
  ) {
    throw new ApiClientError("后端 Protobuf 解码响应格式不正确。", {
      code: "INVALID_RESPONSE",
    });
  }
  return response as ProtoDecodeResult;
}
